### PR TITLE
[#750] PoC of HealthCheckServer that gets health pushed by verticles.

### PR DESCRIPTION
**Pull request by b-abel**
This is a proposal for a solution for #750 which provides an alternative to PR #873.
This is not a complete implementation, **it is just a draft** to see if we like the direction in which it is going and so that can discuss variants of it. I hope you are OK with that.

The basic idea is that verticles actively report their health status to the `HealthCheckServer` over the eventbus. Respectively no reporting for a longer time in the case of the liveness checks could be seen as a failing check because we can not expect that the verticle will be able to send a message if it is no longer live.
The main advantages over the solution in PR #873 are that the verticles don't need to hold an object reference to `HealthCheckServer` and that the health checks on the verticles side could be implemented in a "event-driven" style.

At first I kept the results of the `Status` objects in maps but then I realized that Vert.x's `HealthCheckHandler` already does the same thing, so I simple re-register a check when a status is updated.

What is missing:

- Health check of Endpoints
- readiness checks in applications that are not protocol adapters 
- readiness updates in `AbstractProtocolAdapterBase` when connection to a service is lost
- error handling, parameter checks, JavaDoc, refactoring ...
- Removal of the old mechanism (`HealthCheckProvider`)
- cleanup and write tests
- I had to increase the property `service.startup.timeout` in the integration tests for the tests to complete (at least on my machine)

BTW: I found the comment "//TODO Event Bus support" by accident in the class `HealthCheckHandlerImpl` and wonder if something similar is planned in Vert.x.

### CODE REVIEW

File: service-base/src/main/java/org/eclipse/hono/service/HealthCheckServer.java [hunk 1]
```diff
@@ -87,6 +107,60 @@ private void prepareHealthCheck() {
...
+        final String instanceId = msg.headers().get(MSG_FIELD_VERTICLE_INSTANCE_ID);
+        final String procedureName = msg.headers().get(MSG_FIELD_PROCEDURE_NAME);
+        final Status status = new Status(msg.body());
+
+        LOG.info("received readiness update {}/{}/{}: {}", verticleName, instanceId, procedureName, status.isOk());
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("readiness update status: {}", msg.body().encodePrettily());
+        }
+
+        readinessHandler.register(verticleName + "/" + instanceId + "/" + procedureName,
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/HealthCheckServer.java]: I am not sure if this method is supposed to be called *dynamically* during run time. My understanding is that it should be invoked during startup to set up the checks once. My concern is that this method is quite costly to invoke but I haven't checked.

File: service-base/src/main/java/org/eclipse/hono/service/HealthCheckServer.java [hunk 2]
```diff
@@ -87,6 +107,60 @@ private void prepareHealthCheck() {
...
+    }
+
+    private void processReadinessMessage(final Message<JsonObject> msg) {
+
+        final String verticleName = msg.headers().get(MSG_FIELD_VERTICLE_NAME);
+        final String instanceId = msg.headers().get(MSG_FIELD_VERTICLE_INSTANCE_ID);
+        final String procedureName = msg.headers().get(MSG_FIELD_PROCEDURE_NAME);
+        final Status status = new Status(msg.body());
+
+        LOG.info("received readiness update {}/{}/{}: {}", verticleName, instanceId, procedureName, status.isOk());
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/HealthCheckServer.java]: IMHO DEBUG should be enough. in the end, the readiness will be monitored by an external system that uses this information. So, if anybody is interested in getting notified about a change, they will probably do so based on the external monitoring system, right?

File: tests/pom.xml [hunk 3]
```diff
@@ -33,7 +33,7 @@ Test cases are run against Docker images of Hono server + (Apache Qpid Dispatch
     <maven.source.skip>true</maven.source.skip>
     <maven.install.skip>true</maven.install.skip>
     <gpg.skip>true</gpg.skip>
-    <service.startup.timeout>120000</service.startup.timeout>
+    <service.startup.timeout>160000</service.startup.timeout>
```
Comment (sophokles73) [inline tests/pom.xml]: this doesn't seem to be related to this PR, is it?

**Comment by sophokles73** [comment]
I generally like this approach better than #873 because it better decouples the verticle instances from the health check server itself.

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java [hunk 4]
```diff
@@ -40,6 +44,9 @@
  */
 public abstract class AbstractServiceBase<T extends ServiceConfigProperties> extends ConfigurationSupportingVerticle<T> implements HealthCheckProvider {
 
+    // TODO make configurable
+    protected static final long HEARTBEAT_RATE_MILIIS = HealthCheckServer.LIVENESS_TIMEOUT_SECONDS * 500;
```
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: The idea is to send updates more frequently than the timeout to prevent small delays from causing failing liveness checks. Do you object this idea or just need the reasoning to be more explicit?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java [hunk 5]
```diff
@@ -75,9 +87,38 @@ public final void setTracer(final Tracer opentracingTracer) {
...
+    protected final void sendHealthStatus(final String procedureName, final Status status, final String address) {
+        final Instant now = Instant.now();
+
+        final DeliveryOptions options = new DeliveryOptions()
+                .addHeader(HealthCheckServer.MSG_FIELD_VERTICLE_NAME, this.getClass().getName())
+                .addHeader(HealthCheckServer.MSG_FIELD_VERTICLE_INSTANCE_ID, instanceID.toString())
+                .addHeader(HealthCheckServer.MSG_FIELD_PROCEDURE_NAME, procedureName)
+                .addHeader(HealthCheckServer.MSG_FIELD_LAST_SEEN, String.valueOf(now.getEpochSecond()));
+
+        status.getData().put(HealthCheckServer.MSG_FIELD_LAST_SEEN, now.toString());
```
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: Yes, I like to have the time as a readable string in the `Status` because I found it very convenient when I manually checked the readiness or liveness probes. The epoch timestamp in the header is evaluated for the liveness check by the health check server. But it is of course redundant and I understand that we want to be resource efficient. So we can drop one of them.

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: I am not sure if I am convinced of the liveness check at all ;-) In the end, there usually is not much to check. If the VM has crashed, then there will be no response to the external liveness check requests anyway. The only reasonable liveness check that we have implemened so far is one for checking if the event loop is blocked. But in that case the test itself runs for several seconds (waiting for the timeout to occur) so in that case it doesn't seem to make much sense to set the liveness timeout to say 2 secs and run the check every second, does it?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java [hunk 6]
```diff
@@ -483,16 +485,24 @@ protected void doStop(final Future<Void> stopFuture) {
         final Handler<ProtonConnection> disconnectHandler = getHandlerForDisconnectHonoService(client, serviceName,
                 connectionEstablishedHandler, connectionLostHandler);
 
+        updateReadiness(procedureNameConnectedTo(serviceName), Status.KO()); // initially not ready
```
[the remark this answers is not in the document]
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: I agree, I was only wondering if we need to explicitly report the KO value in order for this to be the case? Wouldn't the readiness check fail anyways if no result has been reported yet?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: > But it is of course redundant and I understand that we want to be resource efficient. So we can drop one of them.

I think we can afford the few extra bytes. The *readable* last seen will then also be reported to the external monitoring tool as part of the HTTP response body, right?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java [hunk 7]
```diff
@@ -75,9 +87,38 @@ public final void setTracer(final Tracer opentracingTracer) {
...
+
+        if (vertx != null) {
+            vertx.setPeriodic(HEARTBEAT_RATE_MILIIS, id -> updateLiveness("heartbeat", Status.OK()));
+        }
+
         startInternal().setHandler(startFuture.completer());
     }
 
+    protected void updateReadiness(final String procedureName, final Status status) {
+        sendHealthStatus(procedureName, status, HealthCheckServer.EVENT_BUS_ADDRESS_HEALTH_READINESS);
```
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: Sorry, I don't understand what you mean.

[the remark this answers is not in the document]
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: I do not see why you would need to pass the address in to the sendHealthStatus method, given that the address is always the same  ...

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: > Wouldn't the readiness check fail anyways if no result has been reported yet?

No, not with the way the checks are used here. The mechanism implemented in this PR is that verticles perform checks and send the results via `EventBus` to `HealthCheckServer` which then registers a `procedure` (which essentially is a name and a function) at a Vert.X `HealthCheckHandler`. (Note that in our case the check function does not perform a check, but just returns the outcome of the check reported).
The consequence is, that `HealthCheckServer` does not know about the check and does not register a procure until it receives the first "update". 
In other words there is no difference between "registering" a check and "updating" it.
I hope this explanation is helpful and not even more confusing :-)

[the remark this answers is not in the document]
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: So, you are saying that the health check for a service succeeds until reported otherwise by any of the verticles etc?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: The proposal for the liveness check in this PR is to send a "heartbeat" from each verticle instance and if it stops arriving in the expected time frame at the `HealthCheckServer` the verticle instance is assumed to be "dead".  As I understand it, this could replace the current check for a blocked event loop. 
We could take the value for the maximal time frame after a verticle instance is considered "dead" for example from the Kubernetes configuration of `livenessProbe.periodSeconds` (which defaults to 10 seconds). The heartbeat needs to be sent more frequently than this maximal time frame to prevent that a small delay, e.g. on the event bus, triggers a failing liveness check.
The idea of the heartbeat is optional. Alternatively we could regularly schedule the current check for a blocked event loop.

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: The `sendHealthStatus` method is used by the method `updateReadiness` and by `updateLiveness`, which indeed only differ by the address they send to.

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: Yes.

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: Yes, that is probably the shortest way to put it :-)
When no check is registered on a `HealthCheckHandler` the liveness probe (e.g. http://127.0.0.1:8088/readiness) returns `204 No Content` and for Kubernetes the HTTP codes >=200 < 400 indicate success.

File: service-base/src/main/java/org/eclipse/hono/service/HealthCheckServer.java (hunk 1 above)
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/HealthCheckServer.java]: It is probably not supposed to be used that way. But on the other hand I saw no problem doing so. It contains a synchronized block which should be no problem since we have only one instance of the class using it. I expect it to be not more expensive than managing the data structures for all results ourselves, whereas the event bus usage should be much more costly.

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: Hm, lets try to walk through this: assume that the service is deployed to Kubernetes and run in a pod with a readiness check configured. When the pod starts and Kubernetes invokes the readiness check URL for the first time, then the result may or may not reflect the real status of the pod because it will depend on whether all verticles have already reported their initial `KO` status or not, right? If they have, then the result of the check will reflect the real status (KO), however, if at least one verticle hasn't, then the result might be correct or not, depending on whether the *missing* verticles are already OK or KO.

The problem seems to be that the health check server has no clue of how many verticles will be  contributing to the overall status, right?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractServiceBase.java]: I see

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (b-abel) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: You are right. Since we are already digging deep: there is the additional option in Kubernetes to configure `initialDelaySeconds` on a health check, which is probably for such cases (if a `readinessProbe` is configured, the ["default state of readiness before the initial delay is `Failure`"](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes)).

> The problem seems to be that the health check server has no clue of how many verticles will be contributing to the overall status, right?

Exactly. The price for decoupling the health check server from the verticals is, that the former one does not know which verticles exist, how many instances have been deployed and which checks (procedures) they perform, _until_ they have reported them for the first time.

Regarding the concrete piece of code here, I still think it correct from the point of view of the Protocol Adapter to start reporting that it is not ready as early as possible and then change the state to ready as soon as it is connected to the required services.

**Comment by b-abel** [comment]
Attached an [example readiness check output](https://github.com/eclipse/hono/files/2639440/example_readiness_check_output.txt) with 2 instances of `VertxBasedHttpProtocolAdapter`. It has 3 levels of checks, each with an id and an aggregated status of it's child checks: 

1. Verticle ("org.eclipse.hono.adapter.http.vertx.VertxBasedHttpProtocolAdapter")
2. Verticle instance (instance UUID)
3. performed check (here state of connection to a service)

**Comment by sophokles73** [comment]
Based on the discussion following up from one of [my comments above](https://github.com/eclipse/hono/pull/910#discussion_r236239681) I no longer think that this is the way to go. FMPOV the result of the health check should be deterministic at all times which seems to be hard to achieve using this approach. Also, the way that external monitoring tools will likely use the health check, i.e. invoke the URL and wait a certain amount of time for the result, does not really match very well with the mechanism of asynchronously reporting the status from within the verticles.

Let's continue with #873  instead.
